clang/include/clang/AST/Type.h

```diff
@@ -4253,7 +4260,7 @@ class FunctionProtoType final
     FunctionType::ExtInfo ExtInfo;
     unsigned Variadic : 1;
     unsigned HasTrailingReturn : 1;
-    unsigned AArch64SMEAttributes : 6;
```

Comment:
Note that thanks to bitfield packing, this is expanding this struct an entire byte.

Reply:
<General Concern>
Looking through the review which started introducing SME attributes, we went to pretty big lengths to try to keep that at zero-cost by figuring out the upper bounds. Now we're looking to expand those upper bounds because they were insufficient less than six months later.

So I'm also concerned about the extra memory pressures this adds because those pressures impact everyone, not just folks targeting AArch64 with these SME attributes. If we have to increase that pressure because this is needed functionality, I think it seems like the cost should be borne by the users using the new functionality instead of impacting everyone.
</General Concern>

That said, `ExtInfo` and `FunctionTypeExtraBitfields` are the ones that needs to stay small; `ExtProtoInfo` is already huge thanks to `ExceptionSpecInfo` being 48 bytes. IIRC, `ExtProtoInfo` is allocated as-needed and so it can grow.

Reply:
Thanks for both your input! The Arm attributes will probably be used reasonably sparsely in the sense that they particularly target code to optimise using SME intrinsics/state. By not adding the attribute it will be a normal function that does not share any SME state. So I agree it makes sense to refactor the code a bit such that these attributes don't affect code that does not use any of these attributes.

I've created #78424 to do this, so that the Arm SME attributes will have their own TrailingObject and are therefore only allocated when a function uses the attributes. I hope that addresses the concerns! (When that patch lands, I'll rebase this one)

It will be more difficult to do something for ExtProtoInfo, given how this is designed, but as @AaronBallman says the struct is already huge and is allocated as-needed. My experiment showed that adding more bits didn't make a practical difference to the size of the ExtProtoInfo. Not sure if this can differ between compilers, but I didn't see it expanding the struct by a byte.